internal/common/auth/authorization/kubernetes.go

```diff
@@ -0,0 +1,203 @@
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/patrickmn/go-cache"
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/G-Research/armada/internal/common/auth/configuration"
+)
+
+type KubernetesNativeAuthService struct {
+	KidMappingFileLocation string
+	TokenCache             *cache.Cache
+	InvalidTokenExpiry     int64
+}
+
+func NewKubernetesNativeAuthService(config configuration.KubernetesAuthConfig) KubernetesNativeAuthService {
+	cache := cache.New(5*time.Minute, 5*time.Minute)
+	return KubernetesNativeAuthService{
+		KidMappingFileLocation: config.KidMappingFileLocation,
+		TokenCache:             cache,
+		InvalidTokenExpiry:     config.InvalidTokenExpiry,
+	}
+}
+
+type CacheData struct {
+	Name  string `json:"name"`
+	Valid bool   `json:"valid"`
+}
+
+func (authService *KubernetesNativeAuthService) Authenticate(ctx context.Context) (Principal, error) {
+	// Retrieve token from context.
+	authHeader := strings.SplitN(metautils.ExtractIncoming(ctx).Get("authorization"), " ", 2)
+
+	if len(authHeader) < 2 || authHeader[0] != "KubernetesAuth" {
+		return nil, missingCredentials
+	}
+
+	token, ca, err := parseAuth(authHeader[1])
+	if err != nil {
+		return nil, missingCredentials
+	}
+
+	// Get token time
+	expirationTime, err := parseTime(token)
+	if err != nil {
+		return nil, err
+	}
+
+	if time.Now().After(*expirationTime) {
+		return nil, fmt.Errorf("invalid token, expired")
+	}
+
+	// Check Cache
+	data, found := authService.TokenCache.Get(token)
+	if found {
+		if cacheInfo, ok := data.(CacheData); ok {
+			if cacheInfo.Valid {
+				return NewStaticPrincipal(cacheInfo.Name, []string{cacheInfo.Name}), nil
+			} else {
+				return nil, fmt.Errorf("token invalid")
+			}
+		}
+	}
+
+	// Get URL from token KID
+	url, err := authService.getClusterURL(token)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make request to token review endpoint
+	name, err := authService.reviewToken(ctx, url, token, []byte(ca))
+	if err != nil {
+		return nil, err
+	}
+
+	// Add to cache
+	authService.TokenCache.Set(
+		token,
+		CacheData{
+			Name:  name,
+			Valid: true,
+		},
+		expirationTime.Sub(time.Now()))
+
+	// Return very basic Principal
+	return NewStaticPrincipal(name, []string{name}), nil
+}
+
+func (authService *KubernetesNativeAuthService) getClusterURL(token string) (string, error) {
+	header := strings.Split(token, ".")[0]
+	decoded, err := base64.RawURLEncoding.DecodeString(header)
+	if err != nil {
+		return "", err
+	}
+
+	var unmarshalled struct {
+		Kid string `json:"kid"`
+	}
+
+	if err := json.Unmarshal(decoded, &unmarshalled); err != nil {
+		return "", err
+	}
+
+	if unmarshalled.Kid == "" {
+		return "", fmt.Errorf("kubernetes serviceaccount token KID must not be empty")
+	}
+
+	url, err := os.ReadFile(authService.KidMappingFileLocation + unmarshalled.Kid)
```

Comment:
this is probably overkill, but could someone trick you into reading random files on the filesystem by sending a `kid` of the form `../../path/to/random/file`

Reply:
I see what you're saying, do you feel there should be some validation checking on the string format of the kid? ensure it doesn't contain slashes of any form?